### ISSUE TITLE
ply fails on python-2.6

### ISSUE THREAD

**Issue by hundeboll** [oe-lite/core#27]
Running OE-lite on an ancient Redhat machine generates the following error:

```
ERROR: exception in bake.run()

Traceback (most recent call last):
  File "/mnt/tmp/buildbot/slave/maquet-deploy/build/bakery/oebakery/cmd/cmds.py", line 137, in call
    ret = function(*args)
  File "/mnt/tmp/buildbot/slave/maquet-deploy/build/oe-lite/meta/core/lib/oelite/cmd/bake.py", line 29, in run
    baker = oelite.baker.OEliteBaker(options, args, config)
  File "/mnt/tmp/buildbot/slave/maquet-deploy/build/oe-lite/meta/core/lib/oelite/baker.py", line 115, in __init__
    self.confparser = confparse.ConfParser(self.config)
  File "/mnt/tmp/buildbot/slave/maquet-deploy/build/oe-lite/meta/core/lib/oelite/parse/confparse.py", line 10, in __init__
    super(ConfParser, self).__init__(data, parent, **kwargs)
  File "/mnt/tmp/buildbot/slave/maquet-deploy/build/oe-lite/meta/core/lib/oelite/parse/oeparse.py", line 18, in __init__
    self.tokens = lexer.lextokens.keys()
AttributeError: 'set' object has no attribute 'keys'

CRITICAL: bake failed: Exception: 'set' object has no attribute 'keys'
```

I am no expert in python-ply, but let me know if I can do any testing.

```
$ python --version
Python 2.6.6
```

**Comment by hundeboll** [comment]
After digging around, it turns out I had ply-3.6 installed, while oelite/core works only with ply-3.4... Downgrading removed the issue.

**Comment by esben** [comment]
Ok.

Although I cannot point at something concrete, I suspect that you might run into some problems using Python 2.6 (or older) with current OE-lite/core.

**Comment by esben** [comment]
Also, FWIW: From changelog of PLY 3.5:

04/16/15: beazley
          Support for versions of Python <2.7 is officially dropped.  PLY may work, but
          the unit tests requires Python 2.7 or newer.

So you might not want to use PLY 3.6 with Python 2.6 anyway.

**Comment by mikaelj** [comment]
Python-2.7 on PLY-3.6.

Which Python versions are supported for PLY-3.6?

**Comment by kimrhh** [comment]
for now you need to downgrade ply, I don't know if this is a bug in oe-lite or ply. We should of course look into fixing it, but personally I can't allocate time for it right now.

**Comment by mikaelj** [comment]
Yup, that's what we've done. Any idea right now what a simple fix is, in
the meantime, or does it require more investigation? The code is a bit too
unfamiliar to me to get it right away

- M

**Comment by kimrhh** [comment]
https://github.com/dabeaz/ply/commit/c34d9d515a2a27ac5d07cb9da1e12b8bd56e6fd0 line 940. They changed the variable to be a set instead of a dict. It smells like we are perhaps using some internals that we should not mess with. Esben?

**Comment by esben** [comment]
That sounds plausible.  We might get away with checking if it is a set or dict, and then handle it accordingly.

**Comment by mikaelj** [comment]
```git

diff --git a/lib/oelite/parse/oeparse.py b/lib/oelite/parse/oeparse.py
index f519080..75e6c2b 100644
--- a/lib/oelite/parse/oeparse.py
+++ b/lib/oelite/parse/oeparse.py
@@ -15,7 +15,12 @@ class OEParser(object):
             lexer = oelite.parse.oelexer
         self.lexer = lexer.clone()
         self.lexer.parser = self
-        self.tokens = lexer.lextokens.keys()
+        if type(lexer.lextokens) == set:
+            # ply >= 3.6
+            self.tokens = list(lexer.lextokens)
+        else:
+            # ply <= 3.4
+            self.tokens = lexer.lextokens.keys()
         oelite.util.makedirs("tmp/ply")
         picklefile = "tmp/ply/" + self.__class__.__module__ + ".p"
         self.yacc = ply.yacc.yacc(module=self, debug=0, picklefile=picklefile)
```

**Comment by esben** [comment]
LGTM. Could you you make a PR with that?

**Comment by kimrhh** [comment]
Tracked upstream here: https://github.com/dabeaz/ply/issues/75

**Comment by esben** [comment]
Work around merged with PR #28 (on master) and PR #39 (4.0).
